reconcile/openshift_serviceaccount_tokens.py

```diff
@@ -47,117 +53,147 @@ def get_tokens_for_service_account(
+            )
             oc_resource = construct_sa_token_oc_resource(oc_resource_name, sa_token)
             ri.add_desired(
-                cluster_name, namespace_name, "Secret", oc_resource_name, oc_resource
+                namespace.cluster.name,
+                namespace.name,
+                "Secret",
+                oc_resource_name,
+                oc_resource,
             )
 
 
-def write_outputs_to_vault(vault_path, ri):
+def write_outputs_to_vault(
+    vault_client: VaultClient, vault_path: str, ri: ResourceInventory
+) -> None:
     integration_name = QONTRACT_INTEGRATION.replace("_", "-")
-    vault_client = VaultClient()
     for cluster, namespace, _, data in ri:
         for name, d_item in data["desired"].items():
             body_data = d_item.body["data"]
             # write secret to per-namespace location
             secret_path = (
-                f"{vault_path}/{integration_name}/" + f"{cluster}/{namespace}/{name}"
+                f"{vault_path}/{integration_name}/{cluster}/{namespace}/{name}"
             )
             secret = {"path": secret_path, "data": body_data}
-            vault_client.write(secret)
+            vault_client.write(secret)  # type: ignore
             # write secret to shared-resources location
-            secret_path = (
-                f"{vault_path}/{integration_name}/" + f"shared-resources/{name}"
-            )
+            secret_path = f"{vault_path}/{integration_name}/shared-resources/{name}"
             secret = {"path": secret_path, "data": body_data}
-            vault_client.write(secret)
+            vault_client.write(secret)  # type: ignore
 
 
-def canonicalize_namespaces(namespaces):
+def canonicalize_namespaces(namespaces: Iterable[NamespaceV1]) -> list[NamespaceV1]:
     canonicalized_namespaces = []
-    for namespace_info in namespaces:
-        if ob.is_namespace_deleted(namespace_info):
-            continue
-        ob.aggregate_shared_resources(namespace_info, "openshiftServiceAccountTokens")
-        openshift_serviceaccount_tokens = namespace_info.get(
-            "openshiftServiceAccountTokens"
-        )
-        if openshift_serviceaccount_tokens:
-            canonicalized_namespaces.append(namespace_info)
-            for sat in openshift_serviceaccount_tokens:
-                canonicalized_namespaces.append(sat["namespace"])
+    for namespace in namespaces:
+        ob.aggregate_shared_resources_typed(namespace)
+        if namespace.openshift_service_account_tokens:
+            canonicalized_namespaces.append(namespace)
+            for sat in namespace.openshift_service_account_tokens:
+                if f"{sat.namespace.cluster.name}/{sat.namespace.name}" in [
+                    f"{ns.cluster.name}/{ns.name}" for ns in canonicalized_namespaces
+                ]:
+                    # do not add duplicate namespaces to speed up fetch_current_state
+                    continue
```

Comment:
how about restructure `canonicalized_namespaces` to `dict` to do auto deduplicate like 

```python
    canonicalized_namespaces = {}
    for namespace in namespaces:
        ob.aggregate_shared_resources_typed(namespace)
        if namespace.openshift_service_account_tokens:
            canonicalized_namespaces[f"{namespace.cluster.name}/{namespace.name}"] = (
                namespace
            )
            for sat in namespace.openshift_service_account_tokens:
                canonicalized_namespaces[
                    f"{sat.namespace.cluster.name}/{sat.namespace.name}"
                ] = NamespaceV1(
                    **sat.namespace.dict(by_alias=True),
                    sharedResources=None,
                    openshiftServiceAccountTokens=None,
                )

    return list(canonicalized_namespaces.values())
```

may need to adjust tests as result order is undefined.

Reply:
Unfortunately, this won't work. It will override already existing namespaces with `openshiftServiceAccountTokens: [...]` with  `openshiftServiceAccountTokens: None`

Reply:
then current code may have duplicates, like 2 namespaces `N1, N2`, `N1` has a service account token in `N2`, `N2` will be added when iterate `N1` service accounts, later it's added again when processing `N2`.

If we want to keep original `namespace` object, only append when not exist, then we should do it in 2 iterations like 

```python
    canonicalized_namespaces = {}
    for namespace in namespaces:
        ob.aggregate_shared_resources_typed(namespace)
        if namespace.openshift_service_account_tokens:
            canonicalized_namespaces[f"{namespace.cluster.name}/{namespace.name}"] = (
                namespace
            )
    for namespace in canonicalized_namespaces.values():
        for sat in namespace.openshift_service_account_tokens:
            key = f"{sat.namespace.cluster.name}/{sat.namespace.name}"
            if key not in canonicalized_namespaces:
                canonicalized_namespaces[key] = NamespaceV1(
                    **sat.namespace.dict(by_alias=True),
                    sharedResources=None,
                    openshiftServiceAccountTokens=None,
                )
    return list(canonicalized_namespaces.values())
```

Reply:
You are right! 🙇 Applied in 4ac709a